Allow optional parameter named 'include'

jaydorsey opened the pull request:
A parameter named include was creating an issue, because even when not
required, it wasn't being rejected because the `respond_to?` check was
always returning `true`. An `example` was responding to the built-in
RSpec method `include`, versus looking for the memoized `let` variable.

This closes #188 

I think I'd ideally like to have checked this another way, but I don't see a way to
differentiate (on an example) whether a method is defined via a `let` or it's a
method that the example responds to

The reason behind this PR is because the `include` keyword as a query parameter
is a jsonapi standard convention, so I can't well change the parameter name and
follow the standard.

There is a workaround that kind-of works, setting a `let(:include) { {} }` before each
test; but that results in a URL with an empty parameter and isn't intuitive, or
correct. 

This PR allows us to generate documentation without having to specify the
optional `include` query parameter

Update: I need to look into this in depth, again. This still breaks the `include` matcher
when you use a `let(:include)`. I think that's an rspec issue though and I'm not sure 
that's something we can overcome

yxhuvud commented:
This would be extremely nice to have considering that `include` is one of the keywords of `json:api`. See https://jsonapi.org/format/#fetching-includes

jaydorsey commented:
I've since discovered there's still an issue with how this all works, and I'm not sure if it's rspec or rswag side of things:

```
let!(:include) { 'foo' } # ?include=foo

run_test! do
  # This include breaks, because it gets redefined
  expect(<parsed response>).to include([1, 2, 3])
end
```

WIth RSpec, the include matcher appears to get overwritten by the `let`, because the `let` is defined at the top level. I can live with this, however.

My work around is to just never use the `include` matcher when I'm doing a test that uses an `include` param (I'll do `eq([1, 2, 3])` for example) but it's not very elegant

This PR will still let you, at least, use `include` as a parameter name so I think it's still a relevant change. Without this PR the specs always break:

Here's an example of one of my parameters with an `include` defined, to document it:

     parameter name: :include, in: :query, type: :string, required: false

And a sample spec (I'm not even using include in the querystring, or testing it):

        context 'default' do
          let!(:task0) { create(:task) }

          run_test! do
            expect(json_response).to match_json_schema('response')
          end
        end

And here's a sample error:

```
  1) tasks /tasks get tasks default returns a 200 response
     Failure/Error: raise InvalidURIError, "bad URI(is not URI?): #{uri.inspect}"

     URI::InvalidURIError:
       bad URI(is not URI?): "http://www.example.com:80/tasks?page[number]=1&page[size]=25&include=#<RSpec::Matchers::BuiltIn::Include:0x00007f97d56ab8b8>"
```

This PR fixes that error

jaydorsey committed:
Allow optional parameter named 'include'

A parameter named include was creating an issue, because even when not
required, it wasn't being rejected because the `respond_to?` check was
always returning `true`. An `example` was responding to the built-in
RSpec method `include`, versus looking for the memoized `let` variable.

jaydorsey commented:
Closing this PR. Someone else can feel free to re-open.

I no longer use rswag and don't have the time to investigate a better solution for this. Hopefully the example I provided will be useful to others.

kwent commented:
This should be reopened this is still an issue !

varg90 commented:
@kwent unfortunately this PR was not merged, but there is another PR https://github.com/rswag/rswag/pull/436
...which is not merged also.
If there is another way to use `parameter :'include', ...` in our rswag request tests please let us know.
